### PR TITLE
Method 2,3: generalize the use of the SOI and nTS

### PR DESCRIPTION
**Pull request by mariadalfonso**
This PR remove hard coded value for Sample Of Interest and Number of Time slices.

Target for next year release (in case HCAL want to reduce the nTS from 10 and re-adjust the sample of interest currently defined as TS=4) 

Expected changes in run1 data for the recHits affected by the QPLL problems
/afs/cern.ch/user/a/abdullin/public/3d_QPLL_HcalRecoParams/QPLL+LBV_channels.txt

### CODE REVIEW

**Comment by perrotta** [comment]
please test

File: RecoLocalCalo/HcalRecAlgos/src/PulseShapeFitOOTPileupCorrection.cc [hunk 1]
```diff
@@ -406,10 +411,10 @@ int PulseShapeFitOOTPileupCorrection::pulseShapeFit(const double * energyArr, co
    bool  fitStatus   = false;
    bool useTriple = false;
 
-   int BX[3] = {4,5,3};
+   int BX[3] = {int(soi),int(soi+1),int(soi-1)};
```
Comment (perrotta) [inline RecoLocalCalo/HcalRecAlgos/src/PulseShapeFitOOTPileupCorrection.cc]: Are these castings needed? I.e. couldn't BX remain a vector of unsigned ints, also in the fit(..) method

Comment (mariadalfonso) [reply RecoLocalCalo/HcalRecAlgos/src/PulseShapeFitOOTPileupCorrection.cc]: fixed int vs unsigned

**Comment by slava77** [comment]
> Reco comparison results: 1151 differences found in the comparisons

all changes are in Run-1 workflows.

this is  in disagreement with this PR description
>No effect is expected in the current local reco.

**Comment by slava77** [comment]
@mariadalfonso 
please clarify on the status of understanding the differences in run-1 reco triggered by changes in this PR.

**Comment by mariadalfonso** [comment]
@slava77 
I'm testing the Run1 workflow locally. 

I'm a bit puzzled since I changed the Method2,3 but the Run1 should use the M0 (and this should be untouched)

From the cmsbot tests:
 HCAL is ok 
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_9_4_X_2017-10-22-0000+20986/23242/140.53_RunHI2011+RunHI2011+RECOHID11+HARVESTDHI/Hcal.html

but jetMET is not 
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_9_4_X_2017-10-22-0000+20986/23242/140.53_RunHI2011+RunHI2011+RECOHID11+HARVESTDHI/JetMET.html

**Comment by slava77** [comment]
On 10/24/17 10:47 AM, mariadalfonso wrote:
> @slava77 <https://github.com/slava77>
> I'm testing the Run1 workflow locally.
> 
> I'm a bit puzzled since I changed the Method2,3 but the Run1 should use
> the M0 (and this should be untouched)
> 

The same reco is supposed to be applied to all data taking periods.
If M0 default is needed in standard reco, one can go back to CMSSW_5_3_X

**Comment by mariadalfonso** [comment]
@slava77 
I have cmsdas problem with any of the run1 workflow
runTheMatrix.py -l 4.22 -j 8 
runTheMatrix.py -l 140.53 -j 8 
runTheMatrix.py -l 1000.0 -j 8 

The following  fail
das_client --limit 0 --query 'file dataset=/MinimumBias/Run2011A-v1/RAW run=165121 site=T2_CH_CERN'

The following  works
das_client --limit 0 --query 'file dataset=/MinimumBias/Run2011A-v1/RAW'

**Comment by slava77** [comment]
On 10/24/17 12:07 PM, mariadalfonso wrote:
> @slava77 <https://github.com/slava77>
> I have cmsdas problem with any of the run1 workflow
> runTheMatrix.py -l 4.22 -j 8
> runTheMatrix.py -l 140.53 -j 8
> runTheMatrix.py -l 1000.0 -j 8
> 
> The following fail
> das_client --limit 0 --query 'file dataset=/MinimumBias/Run2011A-v1/RAW
> run=165121 site=T2_CH_CERN'
> 
> The following works
> das_client --limit 0 --query 'file dataset=/MinimumBias/Run2011A-v1/RAW'

right, some of the manual intervention is necessary sometimes.

**Comment by mariadalfonso** [comment]
@slava77  @abdoulline
I investigated a bit the workflow 1000.0_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT
The "phase1" code is also used for this RECO.

Most of the "sample of interest" I get from the from the HBHEChannelInfo are 4 as the default hardcoded release.
However in a few cases it is set to 3 and the differences spotted by cmsbot are coming from these rechits.
To be investigated.

**Comment by abdoulline** [comment]
Well, there were some intermittent timing alignment issues in Run 1 data.
We've even had an extended M0 reco window for some groups of channels 
suffering from this intermittent timing issues. Those are 3 groups marked 
as "QPLL" ones in
/afs/cern.ch/user/a/abdullin/public/3d_QPLL_HcalRecoParams/QPLL+LBV_channels.txt

I recall that in some rare cases there could have been not only SOI in TS#3 
but even in TS#5...

**Comment by mariadalfonso** [comment]
@abdoulline 
Thanks for the explanation

I can confirm that for the file I was testing
/store/data/Run2011A/MinimumBias/RAW/v1/000/165/121/0699429A-B37F-E011-A57A-0019B9F72D71.root
I can see the SOI=3 for HE and iphi=55,56 accordingly to your QPLL flags.

@slava77 @perrotta 
Let me know how you want to proceed with this PR.
From the HCAL side we understand the changes in the reco, and should go in the direction of improving the M2 since the parameters SOI and nTS can be taken properly for each channel.

**Comment by abdoulline** [comment]
I think this PR improves the hadling of those channels in Run 1 wrt what 
was right before it, by dynamically following displaced SOI...

**Comment by slava77** [comment]
>  Let me know how you want to proceed with this PR.

please update the PR description with some notes about actually expected changes.

**Comment by perrotta** [comment]
please test

File: RecoLocalCalo/HcalRecAlgos/interface/PulseShapeFitOOTPileupCorrection.h [hunk 2]
```diff
@@ -120,14 +121,14 @@ class PulseShapeFitOOTPileupCorrection
     const HcalPulseShapes::Shape* currentPulseShape_=nullptr;
     void setChi2Term( bool isHPD );
 
-    void setPulseShapeTemplate  (const HcalPulseShapes::Shape& ps, bool isHPD);
-    void resetPulseShapeTemplate(const HcalPulseShapes::Shape& ps);
+    void setPulseShapeTemplate  (const HcalPulseShapes::Shape& ps, bool isHPD, unsigned nSamples_);
```
Comment (slava77) [inline RecoLocalCalo/HcalRecAlgos/interface/PulseShapeFitOOTPileupCorrection.h]: please use consistent style for variable names:
- local variables without underscore
- trailing underscore reserved for data members of the class

**Comment by slava77** [comment]
@mariadalfonso @igv4321 @abdoulline @kpedro88 
do we still have a use case for PulseShapeFitOOTPileupCorrection::apply ?

**Comment by slava77** [comment]
> Expected changes in run1 data for the recHits affected by the QPLL problems
/afs/cern.ch/user/a/abdullin/public/3d_QPLL_HcalRecoParams/QPLL+LBV_channels.txt

>  From the HCAL side we understand the changes in the reco, and should go in the direction of improving the M2 since the parameters SOI and nTS can be taken properly for each channel.

Could you please confirm that this is an improvement.

I see generally a reduction of energy above threshold, which looks like a loss of data.

IIUC, the channels with QPLL issues are **sometimes** shifted off by one TS and M0 reco had larger window to capture signal from both the normal and the offset TS.
The M2 logic however is designed to fit an integral in the SOI only. It looks like in the tests made in cms-bot the reduction in energy means that the data in these channels was actually in the right TS and we are losing energy. A more correct solution here would be to perhaps add the amplitudes from the 3-pulse fit.

**Comment by abdoulline** [comment]
> It looks like in the tests made in cms-bot the reduction in energy means 
> that the data in these channels was actually in the right TS and we are 
> losing energy. A more correct solution here would be to perhaps add the 
> amplitudes from the 3-pulse fit.

Thank you for pin-pointing the issue, Slava.

While admittedly I took desired for real, mis-interpreting the plots 
as an _increase_ after this PR, the phenomenon/issue per se was quite 
complicated, as far as I know. 
It's not really that SOI was (mistakenly) assigned to TS#3, while the 
signal stayed steady in TS#4. Actually there was a "precession" of the 
signal with BXs, so that the signal was moving gradually
(mostly within 25ns range between TS#4 and TS#3) until synchronization was 
re-established periodically (intermittently) and lost again. So that quite 
often the signal in these problematic channels was somehow shared between 
TS#3 and TS#4...  With occasional assigning of SOI to TS#3, while TS#4 still 
contained a big fraction (but I'm not sure if the biggest) of the signal. 
However, I don't immediately see if there is a good simple ("elegant") 
solution for this (wrt old large M0 window)...

**Comment by slava77** [comment]
On 10/27/17 11:05 AM, Salavat Abdullin wrote:
> 
>> It looks like in the tests made in cms-bot the reduction in energy means
>> that the data in these channels was actually in the right TS and we are
>> losing energy. A more correct solution here would be to perhaps add the
>> amplitudes from the 3-pulse fit.
> 
> Thank you for pin-pointing the issue, Slava.
> 
> While admittedly I took desired for real, mis-interpreting the plots
> as an _increase_ after this PR, the phenomenon/issue per se was quite
> complicated, as far as I know.
> It's not really that SOI was (mistakenly) assigned to TS#3, while the
> signal stayed steady in TS#4. Actually there was a "precession" of the
> signal with BXs, so that the signal was moving gradually
> (mostly within 25ns range between TS#4 and TS#3) until synchronization was
> re-established periodically (intermittently) and lost again. So that quite
> often the signal in these problematic channels was somehow shared between
> TS#3 and TS#4... With occasional assigning of SOI to TS#3, while TS#4 still
> contained a big fraction (but I'm not sure if the biggest) of the signal.
> However, I don't immediately see if there is a good simple ("elegant")
> solution for this (wrt old large M0 window)...
> 

I think that the 3-pulse (or even 2-pulse) fit in M2 should be able to
get all of the signal.
Plain addition of the (integrals) of 3 pulses could be an OK solution.
Considering that this period has 50ns bunchspacing the chance to pick up
OOT will be small.
Perhaps something less inclusive could work as well.

**Comment by abdoulline** [comment]
Slava, do you mean to force 3-pulses for all channels for 2011?  But we 
don't have "2011" era, right?
Actually this issue was there till the end of the Run 1, so in 2012 (25ns) 
as well and 3d QPLL issue arrived at the end of 2011 "to join" the first 
two. 
Well, once we've used 4TS M0 reco window in 2011 (50ns) and 3TS 
(TS#3,4,5) one in 2012 (25ns), 3-pulses fit sum might not be that far from 
those... But this might need some investigation/studies, while this PR is 
supposed to be a part of HCAL preparations for 2018 and should be good for 
the entire Run 2.

Can we get back to Run 1 particularities later?

**Comment by slava77** [comment]
Here is some more detailed comparison based on wf 4.53

log10(energy) for hits with differences between the baseline (in black) and this PR (in red)
hbhereco_mismatch_log10energy_rebin

The 2D scatter plot shows that the results go both ways
hbhereco_mismatch_log10energy_2d_colz


Combined, there is more loss than gain. 
But perhaps this is incidental for this specific range of events and the net effect on a larger sample may end up being more even rotation of hit energies in and out .. although the opposite may be true as well.

**Comment by slava77** [comment]
On 10/27/17 1:35 PM, Salavat Abdullin wrote:
> 
> Slava, do you mean to force 3-pulses for all channels for 2011? But we
> don't have "2011" era, right?

I mean to make a 3-pulse fit **and** pick all 3 pulses from the fit and
add the to the total,
instead of the central pulse as we do it now.

> Actually this issue was there till the end of the Run 1, so in 2012 (25ns)
> as well and 3d QPLL issue arrived at the end of 2011 "to join" the first
> two.

I seem to recall that we had essentially no 25 ns running until 2015.
There was a short run in October of 2012, but it didn't make it to the
physics dataset.


> Well, once we've used 4TS M0 reco window in 2011 (50ns) and 3TS
> (TS#3,4,5) one in 2012 (25ns), 3-pulses fit sum might not be that far from
> those... But this might need some investigation/studies, while this PR is
> supposed to be a part of HCAL preparations for 2018 and should be good for
> the entire Run 2.
> 
> Can we get back to Run 1 particularities later?
> 
> 
> 
> On Fri, 27 Oct 2017, Slava Krutelyov wrote:
> 
>> On 10/27/17 11:05 AM, Salavat Abdullin wrote:
>> >
>> >> It looks like in the tests made in cms-bot the reduction in energy
> means
>> >> that the data in these channels was actually in the right TS and we are
>> >> losing energy. A more correct solution here would be to perhaps add the
>> >> amplitudes from the 3-pulse fit.
>> >
>> > Thank you for pin-pointing the issue, Slava.
>> >
>> > While admittedly I took desired for real, mis-interpreting the plots
>> > as an _increase_ after this PR, the phenomenon/issue per se was quite
>> > complicated, as far as I know.
>> > It's not really that SOI was (mistakenly) assigned to TS#3, while the
>> > signal stayed steady in TS#4. Actually there was a "precession" of the
>> > signal with BXs, so that the signal was moving gradually
>> > (mostly within 25ns range between TS#4 and TS#3) until
> synchronization was
>> > re-established periodically (intermittently) and lost again. So that
> quite
>> > often the signal in these problematic channels was somehow shared
> between
>> > TS#3 and TS#4... With occasional assigning of SOI to TS#3, while
> TS#4 still
>> > contained a big fraction (but I'm not sure if the biggest) of the
> signal.
>> > However, I don't immediately see if there is a good simple ("elegant")
>> > solution for this (wrt old large M0 window)...
>> >
>>
>> I think that the 3-pulse (or even 2-pulse) fit in M2 should be able to
>> get all of the signal.
>> Plain addition of the (integrals) of 3 pulses could be an OK solution.
>> Considering that this period has 50ns bunchspacing the chance to pick up
>> OOT will be small.
>> Perhaps something less inclusive could work as well.
>>

**Comment by abdoulline** [comment]
Yes, I meant 3-pulse fit sum as weel (in my second paragraph), but first 
forcing it to do 3-pulse fit all the time. \
You right, we were running with 50ns injection in 2012...

**Comment by slava77** [comment]
related to https://github.com/cms-sw/cmssw/pull/20986#issuecomment-339687522
> @mariadalfonso @igv4321 @abdoulline @kpedro88
do we still have a use case for PulseShapeFitOOTPileupCorrection::apply ?

It looks like this method is not used anymore
I created issue #21083 to possibly clean up.

**Comment by igv4321** [comment]
Maria should probably comment on this. The consequence
of PulseShapeFitOOTPileupCorrection::apply removal will be
our inability to use HcalSimpleRecAlgo and, consequently,
HcalSimpleReconstructor for HBHE reconstruction.

In principle, a more efficient way to approach this kind of
cleanup would be to adapt HBHEPhase1Reconstructor for
reconstructing HO and then get rid of both
HcalSimpleReconstructor and HcalSimpleRecAlgo.
But this step needs a wider discussion among HCAL
experts and HO users.

**Comment by slava77** [comment]
On 10/30/17 10:11 PM, Igor Volobouev wrote:
> Maria should probably comment on this. The consequence
> of PulseShapeFitOOTPileupCorrection::apply removal will be
> our inability to use HcalSimpleRecAlgo and, consequently,
> HcalSimpleReconstructor for HBHE reconstruction.

is this an interface constraint?

Before making the github issue I changed in my local test
PulseShapeFitOOTPileupCorrection::apply implementation to start with
assert(0)
and all of the standard workflows passed successfully.
This means that this method is not used at least in these tests.

**Comment by igv4321** [comment]
Not sure what you mean by "interface constraint" (there is
actually such a concept with a well-defined meaning in C#,
but I am not aware of something like that in C++).

Indeed, after switching to HBHEPhase1Reconstructor for
legacy reco, I don't expect that
PulseShapeFitOOTPileupCorrection::apply should be
used anywhere. But, again, it would be good if Maria
confirms.

**Comment by slava77** [comment]
@mariadalfonso 
please let me know if there was a decision made already on what to do with the changes affecting behavior for run-1 data.

**Comment by mariadalfonso** [comment]
@slava77 @perrotta 

fixed the _ in the variable declaration and re-aligned the time also for the soi shift (as spotted by @abdoulline , only for the cases where the soi is moved from the nominal 4)

For the Run1 pathological rechits there is no desire to adapt the M2 algorithm: i.e. " make a 3-pulse fit **and** pick all 3 pulses from the fit and add the to the total". For Run1 re-reco we likely re-run the old releases or use M0 from the new releases.

There is no case for the PulseShapeFitOOTPileupCorrection::apply (HO/HF do not use the M2 and the HB/HE): in principle a lot of parameters in the python config and old code we used in Run1, can be cleaned up. I will take a note on this. We can do this later and separately together with other code's cleanup.

File: RecoLocalCalo/HcalRecAlgos/interface/PulseShapeFitOOTPileupCorrection.h [hunk 3]
```diff
@@ -26,6 +26,7 @@ namespace HcalConst{
    constexpr int nsPerBX = 25;
    constexpr float iniTimeShift = 92.5f;
    constexpr double invertnsPerBx = 0.04;
+   constexpr int TSshift = 4;
```
Comment (slava77) [inline RecoLocalCalo/HcalRecAlgos/interface/PulseShapeFitOOTPileupCorrection.h]: start local variables and data members with lower case letter. Capitalization is reserved for class names.

**Comment by slava77** [comment]
On 11/2/17 4:46 PM, mariadalfonso wrote:
> @slava77 <https://github.com/slava77> @perrotta
> <https://github.com/perrotta>
> 
> fixed the _ in the variable declaration and re-aligned the time also for
> the soi shift (as spotted by @abdoulline <https://github.com/abdoulline>
> , only for the cases where the soi is moved from the nominal 4)

thanks

> 
> For the Run1 pathological rechits there is no desire to adapt the M2
> algorithm: i.e. " make a 3-pulse fit *and* pick all 3 pulses from the
> fit and add the to the total". For Run1 re-reco we likely re-run the old
> releases or use M0 from the new releases.

any sensible rereco will not fit in the old release scope [whatever
"old" means here].
the nominal policy is to have the same reconstruction for all data
unless there are detector changes.
It's not obvious to me that drifting QPLLs will never reappear.

> 
> There is no case for the PulseShapeFitOOTPileupCorrection::apply (HO/HF
> do not use the M2 and the HB/HE): in principle a lot of parameters in
> the python config and old code we used in Run1, can be cleaned up. I
> will take a note on this. We can do this later and separately together
> with other code's cleanup.
> 

Thank you.
I'll ping on the github issue for this at some point in the future.

**Comment by slava77** [comment]
I think that the improvements to the floating QPLL cases can go to a follow up PR
I made #21135 to not forget about it.

**Comment by abdoulline** [comment]
Slava, just one small comment:  QPLL issue has  gone in Run 2, as BE 
electronics has been replaced both in HB and HE.

File: RecoLocalCalo/HcalRecAlgos/interface/PulseShapeFitOOTPileupCorrection.h (hunk 3 above)
Comment (mariadalfonso) [reply RecoLocalCalo/HcalRecAlgos/interface/PulseShapeFitOOTPileupCorrection.h]: done

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/20986	d46adcb
- automated pickup of SOI for M2 reco; originally intended to have no changes in reco
    - the change actually modifies results for some channels in run1, which have drifting phase lock. These channels were not really handled properly either with the old version or with the new one. Eventually their energy reconstruction can be improved, to be followed up in #21135 
- jenkins tests pass and comparisons with baseline show differences only in run1 for regions with unstable channels

The last commits correctly fix the time assignment. E.g. in wf 4.53 https://github.com/cms-sw/cmssw/pull/20986	044735c (left) vs the latest version (right) shows that the bump at negative time is gone
wf4 53_hit_time_he_low

**Comment by davidlange6** [comment]
+1

**Pull request merged**